pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -19,12 +19,172 @@ package kuberuntime
 import (
 	"fmt"
 	"io"
+	"os"
+	"path"
 
+	"github.com/golang/glog"
 	"k8s.io/kubernetes/pkg/api"
+	runtimeApi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 	"k8s.io/kubernetes/pkg/util/term"
 )
 
+// getContainerLogsPath gets log path for container.
+func getContainerLogsPath(containerName, podUID string) string {
+	return path.Join(podLogsRootDirectory, podUID, fmt.Sprintf("%s.log", containerName))
+}
+
+// generateContainerConfig generates container config for kubelet runtime api.
+func (m *kubeGenericRuntimeManager) generateContainerConfig(container *api.Container, pod *api.Pod, restartCount int, podIP string) (*runtimeApi.ContainerConfig, error) {
+	opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, container, podIP)
+	if err != nil {
+		return nil, err
+	}
+
+	_, containerName, cid := buildContainerName(pod, container)
+	command, args := kubecontainer.ExpandContainerCommandAndArgs(container, opts.Envs)
+	containerLogsPath := getContainerLogsPath(containerName, string(pod.UID))
+	podHasSELinuxLabel := pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SELinuxOptions != nil
+	config := &runtimeApi.ContainerConfig{
+		Name:        &containerName,
+		Image:       &runtimeApi.ImageSpec{Image: &container.Image},
+		Command:     command,
+		Args:        args,
+		WorkingDir:  &container.WorkingDir,
+		Labels:      newContainerLabels(container, pod),
+		Annotations: newContainerAnnotations(container, pod, restartCount),
+		Mounts:      makeMounts(cid, opts, container, podHasSELinuxLabel),
+		LogPath:     &containerLogsPath,
+		Stdin:       &container.Stdin,
+		StdinOnce:   &container.StdinOnce,
+		Tty:         &container.TTY,
+	}
+
+	memoryLimit := container.Resources.Limits.Memory().Value()
+	cpuRequest := container.Resources.Requests.Cpu()
+	cpuLimit := container.Resources.Limits.Cpu()
+	var cpuShares int64
+	if cpuRequest.IsZero() && !cpuLimit.IsZero() {
+		cpuShares = milliCPUToShares(cpuLimit.MilliValue())
+	} else {
+		// if cpuRequest.Amount is nil, then milliCPUToShares will return the minimal number
+		// of CPU shares.
+		cpuShares = milliCPUToShares(cpuRequest.MilliValue())
+	}
+	if cpuShares != 0 || memoryLimit != 0 || m.cpuCFSQuota {
```

Comment:
Agreed it's unclear and confusing. That's why I proposed letting kubelet set sane default, instead of relying on how the runtime may interpret it and applies its own default.